File: arrays-hashing/medium/valid-sudoku/sol.cs
// https://leetcode.com/problems/valid-sudoku/description/

public class Solution
{
    public bool IsValidSudoku(char[][] board)
    {
        Dictionary<int, HashSet<char>> row = new Dictionary<int, HashSet<char>>();
        Dictionary<int, HashSet<char>> col = new Dictionary<int, HashSet<char>>();
        Dictionary<string, HashSet<char>> box = new Dictionary<string, HashSet<char>>();

        for (int i = 0; i < board.Length; i++)
        {
            for (int j = 0; j < board[i].Length; j++)
            {
                char v = board[i][j];

                if (v == '.')
                    continue;

                string boxKey = $"{i / 3}{j / 3}";

                if (
                    (row.ContainsKey(i) && row[i].Contains(v))
                    || (col.ContainsKey(j) && col[j].Contains(v))
                    || (box.ContainsKey(boxKey) && box[boxKey].Contains(v))
                )
                {
                    return false;
                }

                if (!row.ContainsKey(i))
                    row[i] = new HashSet<char>();
                if (!col.ContainsKey(j))
                    col[j] = new HashSet<char>();
                if (!box.ContainsKey(boxKey))
                    box[boxKey] = new HashSet<char>();

                row[(char)i].Add(v);
                col[(char)j].Add(v);
                box[boxKey].Add(v);
            }
        }

        return true;
    }
}

File: arrays-hashing/medium/valid-sudoku/sol.py
# https://leetcode.com/problems/valid-sudoku/description/

from typing import List
from collections import defaultdict


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = defaultdict(set)
        col = defaultdict(set)
        box = defaultdict(set)

        for i in range(len(board)):
            for j in range(len(board[i])):
                v = board[i][j]

                if v == ".":
                    continue

                if v in row[i] or v in col[j] or v in box[(i // 3, j // 3)]:
                    return False

                row[i].add(v)
                col[j].add(v)
                box[(i // 3, j // 3)].add(v)

        return True
